### PR TITLE
[JIT] Last two optimizations to Roslyn.

### PR DESCRIPTION
**Pull request by kumpera**
Improve array stores a bit and do a big rework of the VT WB.

### CODE REVIEW

File: mono/metadata/marshal.c [hunk 1]
```diff
@@ -10013,6 +10017,76 @@ get_virtual_stelemref_wrapper (int kind)
 		mono_mb_emit_exception (mb, "ArrayTypeMismatchException", NULL);
 		break;
 	}
+
+	case STELEMREF_CLASS_SMALL_IDEPTH:
+		/*
```
Comment (vargaz) [inline mono/metadata/marshal.c]: I assume this works because the array has a minimum size so we can't index out of it ?

Comment (vargaz) [inline mono/metadata/marshal.c]: Also, would adding a fastpath (element_class == value_class) help ?

Comment (kumpera) [inline mono/metadata/marshal.c]: Yes, it's the same trick we do on the regular casting code. 
MonoClass::supertypes is never smaller than MONO_DEFAULT_SUPERTABLE_SIZE.

Second point, I haven't tried adding such fast path, but from the previous change that removed it from STELEMREF_CLASS, the hit count in not significant enough to justify it.

**Comment by kumpera** [comment]
Pushed a fix for the linux breakage. Let's see if that was it all.

File: mono/metadata/marshal.c (hunk 1 above)
Comment (vargaz) [inline mono/metadata/marshal.c]: Tested it and the hit rate does appear to be pretty low.

**Comment by kumpera** [comment]
All failures look unrelated.

**Pull request merged**